### ISSUE TITLE
Flask cookies leak (server-side) session values

### ISSUE THREAD

**Issue by dolanjs** [freedomofpress/securedrop#204]
The configured secret key, which is needed to use the flask session feature, is not used to encrypt the session values in the cookie, but only to sign them. This means that the cookie can be decoded to show the values in plain text.

PoC:

```
>>> cookie_str = "eJw9zMsOQ0AYQOFXaf61BVobSRcuGbVApC6Z2TSYMYohUYo23r3VRZcnOfneUPSUdZlgoL_hkIMORF0qYpkVVYmSW2aTpZrsOtqTpDMvHCTTS8OxSCb6e9BKrsYSRHiGTYLiMZS3sW9Y9-f8V6j5dVLhFInADlesxsfACU9-jYQvkMBRrBHbUILUVbwIL57NFRKezztXthnnjIJeZu2DSdD2e97uX30cJrZ9AOFDPgE"
>>> zlib.decompress(base64.urlsafe_b64decode(cookie_str+b"="*(-len(cookie_str) % 4)))
```

```
{"codename": {"b":"Z2xhZCBhd2Z1bCBkaW50IG5vZWwgcGF0dHkgYmVudCBhd2FyZSAxOTYw"},
"csrf_token":{" b":"NzQ5NjVhYWFmODQyY2U3OGQ4NjFmNmFmYTU5ZDA1OWI1MTYxMDg1ZQ=="},
"flagged":false,
"logged_in":true}
```

```
>>> base64.b64decode('Z2xhZCBhd2Z1bCBkaW50IG5vZWwgcGF0dHkgYmVudCBhd2FyZSAxOTYw')
'glad awful dint noel patty bent aware 1960'
```

There are two ways to mitigate this leak: One is to implement an own session interface that uses encryption for the (un)serialization and pass it to the app via the app.session_interface config. The second method is, to only store a session id in the cookie and create a server-side session store with files or a database.

Reported by crew53: SD-01-012 Flask cookies leak (server-side) session values

**Comment by fowlslegs** [comment]
I'm working on EtM scheme for client-side session based on AES-CTR-128 and HMAC-SHA-224, since IMO that's the best option provided by our current crypto toolkit, PyCrypto. I'd love to at some point switch our crypto over to libsodium, but that's another story.

_Each boot_ ([currently every 24 hours](https://github.com/freedomofpress/securedrop/pull/805)), a _new_ encryption and authentication key for both the journalist and source app is generated and _stored in memory_. This provides a mechanism to invalidate client-side session, without the need to add additional attack surface to the application server via an in-memory key/val store (we already have Redis server installed because we use Redis workers, but [we plan to deprecate our Redis dependency](https://github.com/freedomofpress/securedrop/issues/1469)). Granted, the client-side session, if it somehow persists after clicking the logout button or quitting TB, is not really invalidated until the server reboots, this is still better than an infinite lifetime.

This also has the effect that if a cookie is obtained from a source's machine (e.g., it somehow persisted in memory, or was swapped out to disk by the OS) after the next server reboot, not only is it invalid for login, but it is also impossible to tie the cookie to any encrypted submissions if the app server is later compromised (or unencrypted submissions if the SVS key is somehow also compromised). It is plausibly deniable the source even visited SD.

Before getting too excited about any of that though, we should recognize two things:

1. As @diracdeltas points out, in a scenario where the cookie might be obtained from memory or swap, it's also likely the codename itself, which appears in the HTML, may be obtained. To address this issue, I'm actually proposing that after #1448 is merged--which [forces sources to retype their codename to login](https://github.com/freedomofpress/securedrop/pull/1448/files#diff-c61fcfb187dc9b29a103e57cb3f1a8c3R41), with a warning that it's important they remember it, and [without the ability to copy and paste it](https://github.com/freedomofpress/securedrop/pull/1448/files#diff-2c95cf028a2b1a9492121233b607065aR420) to (hopefully) drive in that message--we remove codenames from the HTML besides on the 'generate' view.
2. The other thing to keep in mind about plausible deniability is a favorite quote by Matthew Green:

> “Dammit, they used a deniable key exchange protocol” said no Federal prosecutor ever.
--- https://blog.cryptographyengineering.com/2014/07/26/noodling-about-im-protocols/

The implementation will work the standard way: override `flask.app.session_interface` with a custom subclass of `flask.sessions.SessionInterface`.

**Comment by fowlslegs** [comment]
@justintroutman asked: why SHA-224? The answer is that (i) SHA-224 is considered totally safe, and (ii) it gives us more wiggle room than SHA-256 in terms of a potential non-deterministic website fingerprinting defense that might pad client session. 4B room might not turn out to be significant in terms of expanding the number of sites SecureDrop can be morphed* to (and really this is true if you consider Tor cells are padded to 512B), however, we have not yet shown this and my intuition is to go this way because it's costing us nothing in terms of crypto security. This is also why CTR mode was chosen over CBC, which requires padding.

* Read the [ALPaCA paper](https://3tmaadslguc72xc2.onion/) to understand the term _morphing_ here--essentially it's a method of disguising a web page as another web page, or, more accurately speaking, a probabilistically very average appearing (from the view of the passive observer) site. Read this [circuit fingerprinting paper](https://www.usenix.org/system/files/conference/usenixsecurity15/sec15-paper-kwon.pdf) to see why we're so concerned of the potential impact of website fingerprinting attacks on Hidden Services. And, checkout [the work we're doing](https://github.com/freedomofpress/fingerprint-securedrop) to help evaluate and tackle this problem.

**Comment by fowlslegs** [comment]
So it's come to our attention that we need to replace pycrypto, the progress of which is pretty much stalled (e.g., [this year-old vulnerability](https://github.com/dlitz/pycrypto/issues/173) has still not been fixed in a release, even though it's patched in master--note this one does not effect SD). I wrote most of the code that would resolve this issue as I explained in my last comments (you can check it out at https://github.com/freedomofpress/securedrop/blob/safer-sessions/securedrop/session.py), the only major thing I hadn't figured out yet was how best to share counter objects between Flask threads.

Anyway, since this issue is not high-priority, I figure I might postpone finishing up that branch until after we've replaced our crypto library. Ideally, we'd use https://github.com/pyca/pynacl, which bundles `libsodium`, as a replacement, which would allow me to avoid resolving the synchronization problem as I would definitely use CHACHA20/Poly1305 instead of AES-CTR.

So I'm moving this one to "blocked" in our recently made Kanban-style project board for the 0.4 milestone. Check it out if you haven't: https://github.com/freedomofpress/securedrop/projects/1.

**Comment by eloquence** [comment]
Noting that this is still an issue:
1. Grab `ss` cookie
2. `zlib.decompress(base64.urlsafe_b64decode(cookie_str+"="*(-len(cookie_str) % 4)))` -> result contains codename.

Likely low priority given that cookies are cleared out after a SecureDrop session (we even strongly recommend "New Identity" after the session), but still worth addressing.

**Comment by zenmonkeykstop** [comment]
Closed in error by #6403 ref.

**Comment by legoktm** [comment]
> It could be worth encrypting cookies in the future to reduce potential leakage of the codename on the source's computer, but to do so would not be meaningful without also removing the plaintext codename from the reminder message in the DOM

https://github.com/freedomofpress/securedrop/issues/204#issuecomment-260478199 is a rather complicated version of this; I think if we choose to fix it, we can do a much simpler version that specifically addresses the cookie stealing threat without affecting too much else. It wouldn't close off other risks, like the codename being in the HTML response. But if we encrypt the cookie using a derivation(?) of the existing SECRET_KEY, then we ensure that no one else can read the cookie and we don't add any extra server state or UX changes.
